### PR TITLE
Implemented step out feature in debugger

### PR DESCRIPTION
**Pull request by caiismyname**
Release Notes:   Implemented step out feature in debugger

Changes are inspired by 994daf7924e83529c00cb4d88c453c05915de2d6

Started a test/debugger directory for debugger tests. Currently just aggregating test cases in there until I have time to write a debugger specific test runner.

### CODE REVIEW

File: test/debugger/stepout-test1.js [hunk 1]
```diff
@@ -0,0 +1,26 @@
+/*
+Basic test for stepOut
+Nested function calls -- stepOut on a breakpoint at line 23 (return 4 in function d())
+should allow us to step back into c, then b, then a. 
+*/
+
+let a = function() {
```
Comment (yinghuitan) [inline test/debugger/stepout-test1.js]: The indention in the code looks a bit weird. You may want to be consistent with testcase to use 2 spaces indention.

File: test/debugger/stepout-test1.js [hunk 2]
```diff
@@ -0,0 +1,26 @@
+/*
+Basic test for stepOut
```
Comment (yinghuitan) [inline test/debugger/stepout-test1.js]: How does the test work? Do we have any test runner to verify this step out behavior from code? Or it is just a JS code for manual testing? If so, I do not think it makes sense to checkin.

File: test/debugger/stepout-test1.js [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+Nested function calls -- stepOut on a breakpoint at line 23 (return 4 in function d())
```
Comment (yinghuitan) [inline test/debugger/stepout-test1.js]: Also, please test recursive call cases.

File: src/debugger/server/Stepper.js [hunk 4]
```diff
@@ -72,13 +72,32 @@ export class StepOverStepper extends Stepper {
...
+
+export class StepOutStepper extends Stepper {
+  constructor(filePath: string, line: number, column: number, stackSize: number) {
+    super(filePath, line, column);
+    this._startStackSize = stackSize;
+  }
+  _startStackSize: number;
+
+  isComplete(ast: BabelNode, currentStackSize: number): boolean {
+    if (!this.isAstLocationChanged(ast)) return false;
```
Comment (yinghuitan) [inline src/debugger/server/Stepper.js]: I do not think this is correct, source location may be not changing for recursive function call during step out. For example:
```
function foo() {
   if (depth > 10) {
     return;
   }
   foo();     <=== stop here.
}
```
In the above stop line, you issue a step-out, you may land at the same line in the caller frame.

File: src/debugger/server/Stepper.js [hunk 5]
```diff
@@ -72,13 +72,32 @@ export class StepOverStepper extends Stepper {
...
+    if (currentStackSize < this._startStackSize) {
```
Comment (yinghuitan) [inline src/debugger/server/Stepper.js]: The code can be simplified as
```
return currentStackSize < this._startStackSize;
```
The same for step-over code.

File: test/debugger/stepout-test1.js (hunk 2 above)
Comment (caiismyname) [reply test/debugger/stepout-test1.js]: Currently it's just manual -- I was going to accumulate them until I had a chance to write a test runner. Will remove from commits until the runner is made

File: src/debugger/server/Stepper.js (hunk 4 above)
Comment (caiismyname) [reply src/debugger/server/Stepper.js]: I see. I just tested it and it does affect correctness -- I had assumed it was simply an optimizing heuristic. Question: What benefit does this provide StepOver? It seems like its `if (currentStackSize <= this._startStackSize)` check would always be sufficient and correct.

File: test/debugger/stepout-test1.js (hunk 1 above)
Comment (caiismyname) [reply test/debugger/stepout-test1.js]: What do you mean? This test file is 2-space indented, and so are the exists tests.

File: src/debugger/server/Stepper.js (hunk 4 above)
Comment (caiismyname) [reply src/debugger/server/Stepper.js]: Upon further investigation, it seems like this same line causes the same bug in stepOver.

File: test/debugger/stepout-test1.js (hunk 1 above)
Comment (yinghuitan) [reply test/debugger/stepout-test1.js]: If you look at the indention at line 8, 9, 13, 14 etc..., they all have much bigger indention than 2.

File: src/debugger/server/Stepper.js (hunk 4 above)
Comment (yinghuitan) [reply src/debugger/server/Stepper.js]: Yes, we should fix for step-over case.

File: src/debugger/server/Stepper.js [hunk 6]
```diff
@@ -52,33 +56,49 @@ export class Stepper {
...
-    if (!this.isAstLocationChanged(ast)) return false;
-    if (currentStackSize <= this._startStackSize) {
-      // two cases here:
-      // if current stack length < starting stack length, the program must have
-      // hit an exception so this stepper is no longer relevant
-      // if current stack length === starting stack length, the program returned
-      // to the same stack depth, so a step over is complete
-      return true;
-    }
+    // If current stack length < starting stack length, the program must have
```
Comment (yinghuitan) [inline src/debugger/server/Stepper.js]: "must have" => "either"

File: src/debugger/server/Stepper.js [hunk 7]
```diff
@@ -52,33 +56,49 @@ export class Stepper {
...
+    // hit an exception so this stepper is no longer relevant. Or, the program
+    // has stepped out of a function call, back up to the calling function.
+    if (currentStackSize < this._stepStartData.startStackSize) return true;
```
Comment (yinghuitan) [inline src/debugger/server/Stepper.js]: The whole code logic can be merged as:
```
return currentStackSize < this._stepStartData.startStackSize ||
  (currentStackSize === this._stepStartData.startStackSize) && this.isAstLocationChanged(ast));
```

**Pull request closed without merging**